cryodrgn/utils.py

```diff
@@ -5,10 +5,52 @@
 import logging
 from typing import Tuple
 import numpy as np
+import torch
 
 logger = logging.getLogger(__name__)
 
 
+def meshgrid_2d(lo, hi, n, endpoint=False):
+    """
+    Torch-compatible implementation of:
+    np.meshgrid(
+            np.linspace(-0.5, 0.5, D, endpoint=endpoint),
+            np.linspace(-0.5, 0.5, D, endpoint=endpoint),
+        )
+    Torch doesn't support the 'endpoint' argument (always assumed True)
+    and the behavior of torch.meshgrid is different unless the 'indexing' argument is supplied.
+    """
+    if endpoint:
+        values = torch.linspace(lo, hi, n)
+    else:
+        values = torch.linspace(lo, hi, n + 1)[:-1]
+
+    retval = torch.meshgrid(values, values, indexing="xy")
+
+    # Sanity Check - remove in production
+    check = np.meshgrid(
+        np.linspace(lo, hi, n, endpoint=endpoint),
+        np.linspace(lo, hi, n, endpoint=endpoint),
+    )
+    assert np.allclose(retval[0].cpu(), check[0], atol=1e-7)
+    assert np.allclose(retval[1].cpu(), check[1], atol=1e-7)
+    return retval
+
+
+def window_mask(D, in_rad, out_rad):
```

Comment:
Add a docstring?

Also, give typehints for `in_rad` and `out_rad` and assert that `in_rad <= out_rad`.

Reply:
docstring/typehint/assertion added in f23432225c3ae481a6c6ed8bbec4050932e2c0e0